internal/core/src/exec/expression/Expr.h

```diff
@@ -144,7 +144,7 @@ class SegmentExpr : public Expr {
                 bool allow_any_json_cast_type = false)
 
         : Expr(DataType::BOOL, std::move(input), name),
-          segment_(segment),
+          segment_(const_cast<segcore::SegmentInternalInterface*>(segment)),
```

Comment:
to my understanding, methods that may update internal cache can still be marked as const, with the cache data member being marked as mutable. then we should be able to avoid const_cast.